core/src/databases/table_schema.rs

```diff
@@ -0,0 +1,219 @@
+use std::collections::HashMap;
+
+use anyhow::{anyhow, Result};
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum TableSchemaFieldType {
+    Int,
+    Float,
+    Text,
+    Bool,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct TableSchema(HashMap<String, TableSchemaFieldType>);
+
+impl TableSchema {
+    pub fn from_rows(rows: Vec<Value>) -> Result<Self> {
+        let mut schema = HashMap::new();
+
+        for (row_index, row) in rows.iter().enumerate() {
+            let object = row
+                .as_object()
+                .ok_or_else(|| anyhow!("Row {} is not an object", row_index))?;
+
+            for (k, v) in object {
+                if v.is_null() {
+                    continue;
+                }
+
+                let value_type = match v {
+                    Value::Bool(_) => TableSchemaFieldType::Bool,
+                    Value::Number(x) => {
+                        if x.is_i64() {
+                            TableSchemaFieldType::Int
+                        } else {
+                            TableSchemaFieldType::Float
+                        }
+                    }
+                    Value::String(_) | Value::Object(_) | Value::Array(_) => {
+                        TableSchemaFieldType::Text
+                    }
+                    _ => unreachable!(),
+                };
+
+                if let Some(existing_type) = schema.get(k) {
+                    if existing_type != &value_type {
+                        return Err(anyhow!(
+                            "Field {} has conflicting types on row {}: {:?} and {:?}",
+                            k,
+                            row_index,
+                            existing_type,
+                            value_type
+                        ));
+                    }
+                } else {
+                    schema.insert(k.clone(), value_type);
+                }
+            }
+        }
+
+        Ok(Self(schema))
+    }
+
+    pub fn to_sql_string(&self, table_name: &str) -> String {
+        let mut create_table = format!("CREATE TABLE {} (", table_name);
+
+        for (name, field_type) in &self.0 {
+            let sql_type = match field_type {
+                TableSchemaFieldType::Int => "INT",
+                TableSchemaFieldType::Float => "REAL",
+                TableSchemaFieldType::Text => "TEXT",
+                TableSchemaFieldType::Bool => "BOOLEAN",
+            };
+
+            create_table.push_str(&format!("{} {}, ", name, sql_type));
+        }
+
+        // Remove the trailing comma and space, then close the parentheses.
+        let len = create_table.len();
+        create_table.truncate(len - 2);
+        create_table.push_str(");");
+
+        create_table
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
awesome that you wrote tests :+1:  :pray: 
gives much more confidence in the code :) especially for rust laymen like me